File: nodekit-browser/src/types/expressions/interpreter.ts
import type {List, RegisterId, Value} from "../value.ts";
import type {Action} from "../actions";
import type {Expression, LocalVariableName} from "./expressions.ts";

export interface EvlContext {
    graph_registers: Record<RegisterId, Value>,
    local_variables: Record<LocalVariableName, Value>,
    last_action: Action
}


export function evl(
    expression: Expression,
    context: EvlContext,
): Value  {
    switch (expression.op) {
        // =====================
        // Root
        // =====================
        case "reg": {
            if (!(expression.id in context.graph_registers)) {
                throw new Error(`Graph register '${expression.id}' not found`);
            }
            return context.graph_registers[expression.id]
        }
        case "loc": {
            if (!(expression.name in context.local_variables)) {
                throw new Error(`Local variable '${expression.name}' not found`);
            }
            return context.local_variables[expression.name];
        }
        case "la": {
            return context.last_action;
        }
        case "lit": {
            return expression.value;
        }
        // =====================
        // Conditional
        // =====================
        case "if": {
            const condVal = evl(
                expression.cond,
                context,
            );
            if (condVal) {
                return evl(
                    expression.then,
                    context,
                );
            } else {
                return evl(
                    expression.otherwise,
                    context,
                );
            }
        }

        // =====================
        // Boolean logic
        // =====================
        case "not": {
            const v = evl(
                expression.operand,
                context,
            );
            return !v as Boolean;
        }

        case "and": {
            // short-circuit
            for (const arg of expression.args) {
                const v = evl(
                    arg, context,);
                if (typeof v !== "boolean") {
                    throw new Error(`and: all args must be boolean, got '${typeof v}'`);
                }
                if (!v) return false;
            }
            return true;
        }

        case "or": {
            // short-circuit
            for (const arg of expression.args) {
                const v = evl(arg, context,);
                if (typeof v !== "boolean") {
                    throw new Error(`or: all args must be boolean, got '${typeof v}'`);
                }
                if (v) return true;
            }
            return false;
        }

        // =====================
        // Comparators
        // =====================

        case "eq": {
            const lhs = evl(
                expression.lhs,
                context,
            );
            const rhs = evl(
                expression.rhs,
                context,
            );
            // strict equality; you can swap this for deep equality if needed
            return lhs === rhs;
        }

        case "ne": {
            const lhs = evl(
                expression.lhs,
                context,
            );
            const rhs = evl(
                expression.rhs,
                context,
            );
            return lhs !== rhs;
        }

        case "gt":
        case "ge":
        case "lt":
        case "le": {
            const lhs = evl(
                expression.lhs,
                context,
            );
            const rhs = evl(
                expression.rhs,
                context,
            );

            if (typeof lhs !== typeof rhs) {
                throw new Error(
                    `${expression.op}: lhs and rhs must have same type, got '${typeof lhs}' and '${typeof rhs}'`
                );
            }
            if (typeof lhs !== "number" && typeof lhs !== "string") {
                throw new Error(
                    `${expression.op}: only number or string comparison supported, got '${typeof lhs}'`
                );
            }

            switch (expression.op) {
                case "gt":
                    return (lhs as any) > (rhs as any);
                case "ge":
                    return (lhs as any) >= (rhs as any);
                case "lt":
                    return (lhs as any) < (rhs as any);
                case "le":
                    return (lhs as any) <= (rhs as any);
            }
        }

        // =====================
        // Arithmetic
        // =====================

        case "add":
        case "sub":
        case "mul":
        case "div": {
            const lhs = evl(
                expression.lhs,
                context,
            );
            const rhs = evl(
                expression.rhs,
                context,
            );

            if (typeof lhs !== "number" || typeof rhs !== "number") {
                throw new Error(
                    `${expression.op}: operands must be numbers, got '${typeof lhs}' and '${typeof rhs}'`
                );
            }

            switch (expression.op) {
                case "add":
                    return lhs + rhs;
                case "sub":
                    return lhs - rhs;
                case "mul":
                    return lhs * rhs;
                case "div":
                    if (rhs === 0) {
                        throw new Error("div: division by zero");
                    }
                    return lhs / rhs;
            }
        }

        // =====================
        // Array ops
        // =====================
        case "slice": {
            const arrayVal = evl(
                expression.array,
                context,
            );

            if (!Array.isArray(arrayVal)) {
                throw new Error(`slice: array must be array, got '${typeof arrayVal}'`);
            }

            const startVal = evl(
                expression.start,
                context,
            );
            if (typeof startVal !== "number") {
                throw new Error(`slice: start must be number, got '${typeof startVal}'`);
            }

            let endVal: number | undefined;
            if (expression.end !== null) {
                const evEnd = evl(
                    expression.end,
                    context,
                );
                if (typeof evEnd !== "number") {
                    throw new Error(`slice: end must be number, got '${typeof evEnd}'`);
                }
                endVal = evEnd;
            }
            return arrayVal.slice(startVal, endVal) as List;
        }
        case "map": {
            const arrayVal = evl(
                expression.array,
                context,
            );

            if (!Array.isArray(arrayVal)) {
                throw new Error(`map: array must be array, got '${typeof arrayVal}'`);
            }

            const curName = expression.cur;

            return arrayVal.map((elem) => {
                context.local_variables = {
                    ...context.local_variables,
                    [curName]: elem,
                };
                return evl(
                    expression.func,
                    context,
                );
            }) as List;
        }

        case "filter": {
            const arrayVal = evl(
                expression.array,
                context,
            );
            if (!Array.isArray(arrayVal)) {
                throw new Error(`filter: array must be array, got '${typeof arrayVal}'`);
            }

            const curName = expression.cur;

            const result: List = [];
            for (const elem of arrayVal) {
                context.local_variables = {
                    ...context.local_variables,
                    [curName]: elem,
                };
                const keep = evl(
                    expression.predicate,
                    context,
                );
                if (typeof keep !== "boolean") {
                    throw new Error(
                        `filter: predicate must be boolean, got '${typeof keep}'`
                    );
                }
                if (keep) {
                    result.push(elem);
                }
            }
            return result;
        }

        case "fold": {
            const arrayVal = evl(
                expression.array,
                context,
            );
            if (!Array.isArray(arrayVal)) {
                throw new Error(`fold: array must be array, got '${typeof arrayVal}'`);
            }

            let acc = evl(
                expression.init,
                context,
            );
            const accName = expression.acc;
            const curName = expression.cur;

            for (const elem of arrayVal) {
                context.local_variables = {
                    ...context.local_variables,
                    [accName]: acc,
                    [curName]: elem,
                };
                acc = evl(
                    expression.func,
                    context,
                );
            }
            return acc;
        }

        default: {
            const _exhaustive: never = expression;
            throw new Error(`Unsupported expression op: ${(_exhaustive as any).op}`);
        }
    }
}

File: nodekit/_internal/types/expressions/interpreter.py
from __future__ import annotations

from typing import Dict

from nodekit._internal.types.actions.actions import Action
from nodekit._internal.types.expressions.expressions import (
    Expression,
    LocalVariableName,
)
from nodekit._internal.types.value import List, RegisterId, Value
import pydantic


class EvalContext(pydantic.BaseModel):
    graph_registers: Dict[RegisterId, Value]
    local_variables: Dict[LocalVariableName, Value]
    last_action: Action


def _is_number(x: Value) -> bool:
    # Avoid treating bool as a number, even though bool is a subclass of int
    return isinstance(x, (int, float)) and not isinstance(x, bool)


def evl(expression: Expression, context: EvalContext) -> Value:
    op = expression.op

    # =====================
    # Root
    # =====================
    if op == "reg":
        register_id = expression.id
        if register_id not in context.graph_registers:
            raise KeyError(f"Graph register '{register_id}' not found")
        return context.graph_registers[register_id]

    if op == "local":
        name = expression.name
        if name not in context.local_variables:
            raise KeyError(f"Local variable '{name}' not found")
        return context.local_variables[name]
    if op == "la":
        return context.last_action
    if op == "gli":
        list_val = evl(expression.list, context)
        if not isinstance(list_val, list):
            raise TypeError(f"gli: list must be list, got '{type(list_val).__name__}'")
        index_val = evl(expression.index, context)
        if not isinstance(index_val, int):
            raise TypeError(f"gli: index must be int, got '{type(index_val).__name__}'")
        if index_val < 0 or index_val >= len(list_val):
            raise IndexError(f"gli: index '{index_val}' out of bounds")
        return list_val[index_val]
    if op == "gdv":
        dict_val = evl(expression.dict, context)
        if not isinstance(dict_val, dict):
            raise TypeError(f"gdv: dict must be dict, got '{type(dict_val).__name__}'")
        key_val = evl(expression.key, context)
        if not isinstance(key_val, str):
            raise TypeError(f"gdv: key must be str, got '{type(key_val).__name__}'")
        if key_val not in dict_val:
            raise KeyError(f"gdv: key '{key_val}' not found")
        return dict_val[key_val]
    if op == "lit":
        return expression.value

    # =====================
    # Conditional
    # =====================
    if op == "if":
        cond_val = evl(expression.cond, context)
        if not isinstance(cond_val, bool):
            raise TypeError(
                f"if: condition must be bool, got '{type(cond_val).__name__}'"
            )
        branch = expression.then if cond_val else expression.otherwise
        return evl(branch, context)

    # =====================
    # Boolean logic
    # =====================
    if op == "not":
        v = evl(expression.operand, context)
        if not isinstance(v, bool):
            raise TypeError(f"not: operand must be bool, got '{type(v).__name__}'")
        return not v

    if op == "and":
        for arg in expression.args:
            v = evl(arg, context)
            if not isinstance(v, bool):
                raise TypeError(f"and: all args must be bool, got '{type(v).__name__}'")
            if not v:
                return False
        return True

    if op == "or":
        for arg in expression.args:
            v = evl(arg, context)
            if not isinstance(v, bool):
                raise TypeError(f"or: all args must be bool, got '{type(v).__name__}'")
            if v:
                return True
        return False

    # =====================
    # Comparators
    # =====================
    if op in {"eq", "ne", "gt", "ge", "lt", "le"}:
        lhs = evl(expression.lhs, context)
        rhs = evl(expression.rhs, context)

        if op == "eq":
            return lhs == rhs
        if op == "ne":
            return lhs != rhs

        if type(lhs) is not type(rhs):
            raise TypeError(
                f"{op}: lhs and rhs must have same type, got '{type(lhs).__name__}' and '{type(rhs).__name__}'"
            )
        if not (_is_number(lhs) or isinstance(lhs, str)):
            raise TypeError(
                f"{op}: only number or string comparison supported, got '{type(lhs).__name__}'"
            )

        if op == "gt":
            return lhs > rhs  # type: ignore[operator]
        if op == "ge":
            return lhs >= rhs  # type: ignore[operator]
        if op == "lt":
            return lhs < rhs  # type: ignore[operator]
        if op == "le":
            return lhs <= rhs  # type: ignore[operator]

    # =====================
    # Arithmetic
    # =====================
    if op in {"add", "sub", "mul", "div"}:
        lhs = evl(expression.lhs, context)
        rhs = evl(expression.rhs, context)

        if not (_is_number(lhs) and _is_number(rhs)):
            raise TypeError(
                f"{op}: operands must be numbers, got '{type(lhs).__name__}' and '{type(rhs).__name__}'"
            )

        lhs_num = float(lhs)
        rhs_num = float(rhs)

        if op == "add":
            return lhs_num + rhs_num
        if op == "sub":
            return lhs_num - rhs_num
        if op == "mul":
            return lhs_num * rhs_num
        if op == "div":
            if rhs_num == 0:
                raise ZeroDivisionError("div: division by zero")
            return lhs_num / rhs_num

    # =====================
    # Array ops
    # =====================
    if op == "append":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"append: array must be list, got '{type(array_val).__name__}'"
            )
        value_val = evl(expression.value, context)
        return [*array_val, value_val]

    if op == "concat":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"concat: array must be list, got '{type(array_val).__name__}'"
            )
        value_val = evl(expression.value, context)
        if not isinstance(value_val, list):
            raise TypeError(
                f"concat: value must be list, got '{type(value_val).__name__}'"
            )
        return [*array_val, *value_val]

    if op == "slice":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"slice: array must be list, got '{type(array_val).__name__}'"
            )

        start_val = evl(expression.start, context)
        if not isinstance(start_val, int):
            raise TypeError(
                f"slice: start must be int, got '{type(start_val).__name__}'"
            )

        if expression.end is None:
            end_index = None
        else:
            end_val = evl(expression.end, context)
            if not isinstance(end_val, int):
                raise TypeError(
                    f"slice: end must be int, got '{type(end_val).__name__}'"
                )
            end_index = end_val

        return array_val[start_val:end_index]

    if op == "map":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"map: array must be list, got '{type(array_val).__name__}'"
            )

        cur_name = expression.cur
        base_locals = context.local_variables

        return [
            evl(
                expression.func,
                context.model_copy(
                    update={
                        "local_variables": {**base_locals, cur_name: elem},
                    }
                ),
            )
            for elem in array_val
        ]

    if op == "filter":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"filter: array must be list, got '{type(array_val).__name__}'"
            )

        cur_name = expression.cur
        base_locals = context.local_variables

        result: List = []
        for elem in array_val:
            keep = evl(
                expression.predicate,
                context.model_copy(
                    update={
                        "local_variables": {**base_locals, cur_name: elem},
                    }
                ),
            )
            if not isinstance(keep, bool):
                raise TypeError(
                    f"filter: predicate must be bool, got '{type(keep).__name__}'"
                )
            if keep:
                result.append(elem)
        return result

    if op == "fold":
        array_val = evl(expression.array, context)
        if not isinstance(array_val, list):
            raise TypeError(
                f"fold: array must be list, got '{type(array_val).__name__}'"
            )

        acc = evl(expression.init, context)
        acc_name = expression.acc
        cur_name = expression.cur
        base_locals = context.local_variables

        for elem in array_val:
            acc = evl(
                expression.func,
                context.model_copy(
                    update={
                        "local_variables": {
                            **base_locals,
                            acc_name: acc,
                            cur_name: elem,
                        },
                    }
                ),
            )
        return acc

    raise ValueError(f"Unsupported expression op: {op!r}")
